File: EmarsysPredictShop/EmarsysPredictShop/CategoryViewController.m
//
//  CategoryViewController.m
//  EmarsysPredictShop
//

@import EmarsysPredictSDK;
@import Foundation;

#import "CategoryViewController.h"
#import "ItemDetailViewController.h"
#import "Cart+EmarsysPredictSDKExtensions.h"
#import "DataSource.h"
#import "RecommendedItem.h"

@implementation CategoryViewController

static NSString *const reuseIdentifier = @"ItemCell";

- (void)viewDidLoad {
    [super viewDidLoad];

    self.recommendationManager = [[RecommendationManager alloc] init];
    self.recommendationManager.recommendResults = [[NSMutableArray alloc] init];
    self.itemsManager = [[ItemsManager alloc] init];
    self.itemsManager.items = [[NSArray<Item *> alloc] init];
    self.categoryTableView.dataSource = _itemsManager;
    self.categoryTableView.delegate = _itemsManager;
    self.categorySearchBar.delegate = self;
    self.recommendedCollectionView.dataSource = self.recommendationManager;
    self.recommendedCollectionView.delegate = self.recommendationManager;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    // Deselect all rows
    for (NSIndexPath *indexPath in self.categoryTableView
             .indexPathsForSelectedRows) {
        [self.categoryTableView deselectRowAtIndexPath:indexPath animated:NO];
    }
    self.title = self.category;
    self.originalItems =
        [[DataSource sharedDataSource] itemsFromCategory:self.category];
    self.itemsManager.items = (self.searchTerm && self.searchTerm.length)
                                  ? self.searchResults
                                  : self.originalItems;
    [self sendRecommend];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"ShowItem"]) {
        ItemDetailViewController *ctr = segue.destinationViewController;
        ctr.item =
            self.itemsManager
                .items[self.categoryTableView.indexPathForSelectedRow.row];
    } else if ([segue.identifier isEqualToString:@"ShowRecommendedItem"]) {
        ItemDetailViewController *ctr = segue.destinationViewController;
        RecommendedItem *recItem = sender;
        ctr.item = recItem.item;
    }
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
}

- (void)searchBar:(UISearchBar *)searchBar
    textDidChange:(NSString *)searchText {
    if (searchText && searchText.length) {
        NSPredicate *resultPredicate = [NSPredicate
            predicateWithFormat:@"title BEGINSWITH[c] %@", searchText];
        self.searchResults =
            [self.originalItems filteredArrayUsingPredicate:resultPredicate];
        self.itemsManager.items = self.searchResults;
        self.searchTerm = searchText;
    } else {
        [searchBar resignFirstResponder];
        self.searchTerm = nil;
        self.searchResults = nil;
        self.itemsManager.items = self.originalItems;
    }
    [self.categoryTableView reloadData];
    [self sendRecommend];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)aSearchBar {
    [aSearchBar resignFirstResponder];
}

- (BOOL)searchBar:(UISearchBar *)searchBar
    shouldChangeTextInRange:(NSRange)range
            replacementText:(NSString *)text {

    if ([text isEqualToString:@"\n"]) {
        [searchBar resignFirstResponder];
        return NO;
    }
    return YES;
}

- (void)sendRecommend {

    EMSession *emarsysSession = [EMSession sharedSession];
    EMTransaction *transaction = [[EMTransaction alloc] init];
    NSArray<EMCartItem *> *cartItems = [[Cart sharedCart] convertItems];
    [transaction setCart:cartItems];

    if (self.searchTerm) {
        [transaction setSearchTerm:self.searchTerm];
    }

    [self.recommendationManager.recommendResults removeAllObjects];

    NSString *logic;
    NSMutableArray<NSString *> *exludeItems = nil;

    if (self.searchResults != nil && self.searchResults.count > 0) {
        Item *firstItem = [self.searchResults objectAtIndex:0];
        [transaction setView:firstItem.itemID];

        if (self.searchResults.count > 1) {
            exludeItems = [[NSMutableArray<NSString *> alloc] init];
            int i;
            for (i = 1; i < [self.searchResults count]; i++) {
                Item *item = [self.searchResults objectAtIndex:i];
                [exludeItems addObject:item.itemID];
            }
        }

        logic = @"RELATED";

    } else {
        [transaction setCategory:self.category];
        logic = @"CATEGORY";
    }

    EMRecommendationRequest *recommend =
        [[EMRecommendationRequest alloc] initWithLogic:logic];

    if (exludeItems) {
        [recommend excludeItemsWhere:@"item" isIn:exludeItems];
    }

    recommend.limit = 10;

    recommend.completionHandler = ^(EMRecommendationResult *_Nonnull result) {

      // Process result
      NSLog(@"%@", result.featureID);
      [result.products
          enumerateObjectsUsingBlock:^(EMRecommendationItem *_Nonnull obj,
                                       NSUInteger idx, BOOL *_Nonnull stop) {
            Item *item = [[Item alloc] initWithItem:obj];
            [self.recommendationManager.recommendResults addObject:item];
          }];

      dispatch_async(dispatch_get_main_queue(), ^{
        [self.recommendedCollectionView reloadData];
      });
    };
    [transaction recommend:recommend];

    // Firing the transaction. Should be the last call on the page, called only
    // once.
    [emarsysSession sendTransaction:transaction
                       errorHandler:^(NSError *_Nonnull error) {
                         // Some error happened
                         NSLog(@"value: %@", error);
                         return;
                       }];
}

@end

File: EmarsysMobileShopSwift/EmarsysMobileShopSwift/CategoryViewController.swift
//
//  CategoryViewController.swift
//  EmarsysMobileShopSwift
//
//

import UIKit
import EmarsysMobile

class CategoryViewController: UIViewController, UITabBarDelegate, UISearchBarDelegate {
    
    @IBOutlet weak var recommendedCollectionView: UICollectionView!
    @IBOutlet weak var categoryTableView: UITableView!
    @IBOutlet weak var categorySearchBar: UISearchBar!
    
    var category: String!
    var originalItems = [Item]()
    var searchResults = [Item]()
    var searchTerm: String = ""
    var recommendationManager = RecommendationManager()
    var itemsManager = ItemsManager()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        recommendationManager.recommendResults = [Item]()
        categoryTableView.dataSource = itemsManager
        categoryTableView.delegate = itemsManager
        categorySearchBar.delegate = self
        recommendedCollectionView.dataSource = recommendationManager
        recommendedCollectionView.delegate = recommendationManager
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        title = category
        originalItems = DataSource.sharedDataSource.itemsFromCategory(category)
        itemsManager.items = searchTerm.isEmpty ? originalItems : searchResults
        sendRecommend()
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ShowItem" {
            let ctr = segue.destinationViewController as! ItemDetailViewController
            ctr.item = itemsManager.items[categoryTableView.indexPathForSelectedRow!.row]
        } else if segue.identifier == "ShowRecommendedItem" {
            let ctr = segue.destinationViewController as! ItemDetailViewController
            let recommendedItem = sender as! RecommendedItem
            ctr.item = recommendedItem.item
        }
    }
    
    func searchBar(searchBar: UISearchBar, textDidChange searchText: String) {
        if searchText.isEmpty {
            searchBar.resignFirstResponder()
            searchTerm.removeAll()
            searchResults.removeAll()
            itemsManager.items = originalItems
        } else {
            searchResults = originalItems.filter() {
                if let title = ($0 as Item).title {
                    return title.lowercaseString.rangeOfString(searchText.lowercaseString) != nil
                } else {
                    return false
                }
            }
            itemsManager.items = searchResults
            searchTerm = searchText
        }
        categoryTableView.reloadData()
        sendRecommend()
    }
    
    func searchBarTextDidEndEditing(searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }
    
    func searchBar(searchBar: UISearchBar, shouldChangeTextInRange range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            searchBar.resignFirstResponder()
            return false
        }
        return true
    }
    
    func sendRecommend() {
        let emarsysSession = EMSession.sharedSession()
        let transaction = EMTransaction()
        let cartItems = Cart.sharedCart.convertItems()
        transaction.setCart(cartItems)
        
        if !searchTerm.isEmpty {
            transaction.setSearchTerm(searchTerm)
        }
        
        recommendationManager.recommendResults.removeAll()
        
        let logic: String
        var exludeItems: [String] = [String]()
        
        if searchResults.count > 0 {
            let firstItem = searchResults[0]
            transaction.setView(firstItem.itemID)
            
            if searchResults.count > 1 {
                for item in searchResults {
                    exludeItems.append(item.itemID)
                }
            }
            logic = "RELATED"
            
        } else {
            transaction.setCategory(category)
            logic = "CATEGORY"
        }
        
        let recommend = EMRecommendationRequest(logic: logic)
        
        if exludeItems.count > 0 {
            recommend.excludeItemsWhere("item", isIn: exludeItems)
        }
        
        recommend.limit = 10
        recommend.completionHandler = {result in
            // Process result
            for obj in result.products {
                let item = Item(withItem: obj)
                self.recommendationManager.recommendResults.append(item)
            }
            
            dispatch_async(dispatch_get_main_queue()) {
                self.recommendedCollectionView.reloadData()
            }
        }
        transaction.recommend(recommend)
        
        // Firing the transaction. Should be the last call on the page, called only once.
        emarsysSession.sendTransaction(transaction, errorHandler: {error in
            // Some error happened
            NSLog("value: %@", error)
            return
        })
    }
}
